compiler/rustc_parse/src/parser/ty.rs

```diff
@@ -276,6 +276,34 @@ impl<'a> Parser<'a> {
     ) -> PResult<'a, Box<Ty>> {
         let allow_qpath_recovery = recover_qpath == RecoverQPath::Yes;
         maybe_recover_from_interpolated_ty_qpath!(self, allow_qpath_recovery);
+
+        if (self.token.is_keyword(kw::Const) || self.token.is_keyword(kw::Mut))
+            && self.look_ahead(1, |t| *t == token::Star)
+        {
+            let kw_span = self.token.span;
+            self.bump();
+
+            if self.eat(exp!(Star)) {
```

Comment:
This should try to parse a subsequent type via a snapshot. Or at least check that `lookahead(2)` [can begin a type](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast/token/struct.Token.html#method.can_begin_type)